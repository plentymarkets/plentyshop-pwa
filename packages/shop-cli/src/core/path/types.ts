/**
 * Type definitions for Path Resolution System
 */

/**
 * Options for path resolution
 */
export interface PathOptions {
  /** Custom file name (without extension) */
  fileName?: string;
  /** File extension */
  extension?: string;
  /** Category for settings components */
  category?: string;
  /** Whether page is dynamic */
  isDynamic?: boolean;
  /** Custom web app path */
  webAppPath?: string;
}

/**
 * Result of path resolution
 */
export interface PathResult {
  /** Base directory path (relative to shop-cli) */
  basePath: string;
  /** Main file path */
  mainFile: string;
  /** Types file path (optional) */
  typesFile?: string;
  /** Index file path (optional) */
  indexFile?: string;
  /** Test file path (optional) */
  testFile?: string;
  /** Trigger file path (for settings, optional) */
  triggerFile?: string;
  /** Form file path (for blocks, optional) */
  formFile?: string;
  /** All generated file paths */
  files: string[];
}

/**
 * Strategy interface for path resolution
 */
export interface PathStrategy {
  resolve(name: string, options?: PathOptions): PathResult;
}

/**
 * Configuration interface for path resolution
 */
export interface PathConfig {
  readonly projectRoot: string;
  readonly webAppRoot: string;
  readonly templatesRoot: string;
}
