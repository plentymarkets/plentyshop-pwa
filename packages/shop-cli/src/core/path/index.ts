export * from './PathResolver';
export * from './PathStrategies';
