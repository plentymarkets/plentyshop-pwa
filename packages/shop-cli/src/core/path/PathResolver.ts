/**
 * Path Resolution System for PlentyONE Shop Generators
 *
 * Provides unified path resolution with strategy pattern to replace
 * scattered path logic across template-utils.ts, paths.ts, and ActionBuilder.
 */

import { join, resolve } from 'node:path';
import type { PathOptions, PathResult, PathStrategy, PathConfig } from './types';
import { ComponentPathStrategy, ComposablePathStrategy } from './PathStrategies';

/**
 * Default path configuration
 */
class DefaultPathConfig implements PathConfig {
  get projectRoot(): string {
    // When run from packages/shop-cli, go up two levels to reach project root
    return resolve(process.cwd(), '../../');
  }

  get webAppRoot(): string {
    return join(this.projectRoot, 'apps/web');
  }

  get templatesRoot(): string {
    return join(process.cwd(), 'templates');
  }
}

/**
 * Main path resolver with strategy pattern
 */
export class PathResolver {
  private readonly strategies = new Map<string, PathStrategy>();
  private readonly config: PathConfig;

  constructor(configOverrides?: Partial<PathConfig>) {
    const defaultConfig = new DefaultPathConfig();
    this.config = configOverrides
      ? {
          projectRoot: configOverrides.projectRoot ?? defaultConfig.projectRoot,
          webAppRoot: configOverrides.webAppRoot ?? defaultConfig.webAppRoot,
          templatesRoot: configOverrides.templatesRoot ?? defaultConfig.templatesRoot,
        }
      : defaultConfig;

    this.registerDefaultStrategies();
  }

  /**
   * Register default path strategies
   */
  private registerDefaultStrategies(): void {
    this.strategies.set('component', new ComponentPathStrategy(this.config));
    this.strategies.set('composable', new ComposablePathStrategy(this.config));
  }

  /**
   * Register a custom path strategy
   */
  register(type: string, strategy: PathStrategy): void {
    this.strategies.set(type, strategy);
  }

  /**
   * Resolve paths for a given generator type
   */
  resolve(type: string, name: string, options?: PathOptions): PathResult {
    const strategy = this.strategies.get(type);
    if (!strategy) {
      throw new Error(`No path strategy found for generator type: ${type}`);
    }

    return strategy.resolve(name, options);
  }

  /**
   * Get available generator types
   */
  getAvailableTypes(): string[] {
    return Array.from(this.strategies.keys());
  }

  /**
   * Get template path for a generator type
   */
  getTemplatePath(type: string): string {
    return join(this.config.templatesRoot, type);
  }

  /**
   * Get relative template path for a generator type
   */
  getRelativeTemplatePath(type: string): string {
    return join('templates', type);
  }

  /**
   * Get web app path (for legacy compatibility)
   */
  getWebAppPath(): string {
    return this.config.webAppRoot;
  }

  /**
   * Get project root path
   */
  getProjectRoot(): string {
    return this.config.projectRoot;
  }
}

/**
 * Default path resolver instance
 */
export const pathResolver = new PathResolver();
