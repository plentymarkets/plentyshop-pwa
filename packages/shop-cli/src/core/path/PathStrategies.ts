/**
 * Path strategies for different generator types
 */

import { join } from 'node:path';
import type { PathOptions, PathResult, PathStrategy, PathConfig } from './types';

/**
 * Base strategy for path resolution
 */
abstract class BasePathStrategy implements PathStrategy {
  constructor(protected readonly config: PathConfig) {}

  abstract resolve(name: string, options?: PathOptions): PathResult;

  /**
   * Get relative path from packages/shop-cli to target location
   */
  protected getRelativePath(absolutePath: string): string {
    return join('../../apps/web', absolutePath.replace(this.config.webAppRoot, ''));
  }

  /**
   * Build standard file structure result
   */
  protected buildFileStructure(
    basePath: string,
    name: string,
    options: {
      mainFileExtension: string;
      mainFileName?: string;
      includeTypes?: boolean;
      includeIndex?: boolean;
      includeTests?: boolean;
      testFileName?: string;
    },
  ): PathResult {
    const {
      mainFileExtension,
      mainFileName = name,
      includeTypes = true,
      includeIndex = false,
      includeTests = true,
      testFileName = name,
    } = options;

    const relativePath = this.getRelativePath(basePath);
    const mainFile = `${relativePath}/${mainFileName}.${mainFileExtension}`;

    const result: PathResult = {
      basePath: relativePath,
      mainFile,
      files: [mainFile],
    };

    if (includeTypes) {
      const typesFile = `${relativePath}/types.ts`;
      result.typesFile = typesFile;
      result.files.push(typesFile);
    }

    if (includeIndex) {
      const indexFile = `${relativePath}/index.ts`;
      result.indexFile = indexFile;
      result.files.push(indexFile);
    }

    if (includeTests) {
      const testFile = `${relativePath}/__tests__/${testFileName}.spec.ts`;
      result.testFile = testFile;
      result.files.push(testFile);
    }

    return result;
  }
}

/**
 * Component path strategy
 */
export class ComponentPathStrategy extends BasePathStrategy {
  resolve(name: string, options: PathOptions = {}): PathResult {
    const basePath = join(this.config.webAppRoot, 'components', name);
    const mainFileName = options.fileName || `{{pascalCase name}}`;
    const extension = options.extension || 'vue';

    return this.buildFileStructure(basePath, name, {
      mainFileExtension: extension,
      mainFileName,
      includeTypes: true,
      includeIndex: false,
      includeTests: true,
      testFileName: '{{pascalCase name}}',
    });
  }
}

/**
 * Composable path strategy
 */
export class ComposablePathStrategy extends BasePathStrategy {
  resolve(name: string, options: PathOptions = {}): PathResult {
    const basePath = join(this.config.webAppRoot, 'composables', name);
    const mainFileName = options.fileName || `{{name}}`;
    const extension = options.extension || 'ts';

    return this.buildFileStructure(basePath, name, {
      mainFileExtension: extension,
      mainFileName,
      includeTypes: true,
      includeIndex: true,
      includeTests: true,
      testFileName: '{{name}}',
    });
  }
}
