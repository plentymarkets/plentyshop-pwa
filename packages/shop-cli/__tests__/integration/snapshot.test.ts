/**
 * Snapshot tests for generator file structures
 * Validates that generated files match expected structures
 */

import { describe, it, expect, beforeEach, afterEach } from 'vitest';
import { join } from 'path';
import {
  createSnapshotTestDirectory,
  cleanupTestDirectory,
  createMockComponentStructure,
  createMockComposableStructure,
  generateFileTree,
} from '../utils/index.js';

describe('Generator File Structure Snapshots', () => {
  let testDir: string;
  let testWebAppPath: string;

  beforeEach(async () => {
    const testDirs = await createSnapshotTestDirectory();
    testDir = testDirs.testDir;
    testWebAppPath = testDirs.testWebAppPath;
  });

  afterEach(async () => {
    await cleanupTestDirectory(testDir);
  });

  describe('Component Structure Snapshots', () => {
    it('should match expected component file structure', async () => {
      const componentName = 'ProductCard';
      const componentDir = join(testWebAppPath, 'components', componentName);

      // Create expected component structure
      await createMockComponentStructure(componentDir, componentName);

      const fileTree = await generateFileTree(componentDir);

      expect(fileTree).toMatchSnapshot('component-product-card-structure');
    });
  });

  describe('Composable Structure Snapshots', () => {
    it('should match expected composable file structure', async () => {
      const composableName = 'useProductCart';
      const composableDir = join(testWebAppPath, 'composables', composableName);

      await createMockComposableStructure(composableDir, composableName);

      const fileTree = await generateFileTree(composableDir);
      expect(fileTree).toMatchSnapshot('composable-product-cart-structure');
    });
  });
});
