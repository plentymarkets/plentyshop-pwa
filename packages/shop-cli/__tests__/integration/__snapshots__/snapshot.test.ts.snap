// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Generator File Structure Snapshots > Component Structure Snapshots > should match expected component file structure > component-product-card-structure 1`] = `
{
  "children": [
    {
      "name": "ProductCard.vue",
      "size": 31,
      "type": "file",
    },
    {
      "children": [
        {
          "name": "ProductCard.spec.ts",
          "size": 45,
          "type": "file",
        },
      ],
      "name": "__tests__",
      "type": "directory",
    },
    {
      "name": "types.ts",
      "size": 24,
      "type": "file",
    },
  ],
  "name": "ProductCard",
  "type": "directory",
}
`;

exports[`Generator File Structure Snapshots > Composable Structure Snapshots > should match expected composable file structure > composable-product-cart-structure 1`] = `
{
  "children": [
    {
      "children": [
        {
          "name": "useProductCart.spec.ts",
          "size": 48,
          "type": "file",
        },
      ],
      "name": "__tests__",
      "type": "directory",
    },
    {
      "name": "index.ts",
      "size": 24,
      "type": "file",
    },
    {
      "name": "types.ts",
      "size": 24,
      "type": "file",
    },
    {
      "name": "useProductCart.ts",
      "size": 33,
      "type": "file",
    },
  ],
  "name": "useProductCart",
  "type": "directory",
}
`;
